aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -567,5 +567,260 @@ Tensor& index_add_cuda_(Tensor & self, int64_t dim, const Tensor & index, const
+    indicesInfo.collapseDims();
+
+    // A reasonable choice for when to have each thread iterate over
+    // indices to choose
+    if (numIndices <= 16) {
+      if (outInfo.dims == 1 && selfInfo.dims == 1 && indContig) {
+        SMALL_INDEX(scalar_t, unsigned int, 1, 1, -2);
+      } else if (outInfo.dims == 2 && selfInfo.dims == 2 && indContig) {
+        SMALL_INDEX(scalar_t, unsigned int, 2, 2, -2);
+      } else if (outInfo.dims == 3 && selfInfo.dims == 3 && indContig) {
+        SMALL_INDEX(scalar_t, unsigned int, 3, 3, -2);
+      } else {
+        SMALL_INDEX(scalar_t, unsigned int, -1, -1, -1);
+      }
+    } else {
+      bool indexIsMajor = indexShouldBeMajor(outInfo, outSelectDim);
+
+      if (outInfo.dims == 1 && selfInfo.dims == 1 && indContig) {
+        LARGE_INDEX(scalar_t, unsigned int, 1, 1, -2, true);
+      } else if (outInfo.dims == 2 && selfInfo.dims == 2 && indContig) {
+        if (indexIsMajor) {
+          LARGE_INDEX(scalar_t, unsigned int, 2, 2, -2, true);
+        } else {
+          LARGE_INDEX(scalar_t, unsigned int, 2, 2, -2, false);
+        }
+      } else if (outInfo.dims == 3 && selfInfo.dims == 3 && indContig) {
+        if (indexIsMajor) {
+          LARGE_INDEX(scalar_t, unsigned int, 3, 3, -2, true);
+        } else {
+          LARGE_INDEX(scalar_t, unsigned int, 3, 3, -2, false);
+        }
+      } else {
+        LARGE_INDEX(scalar_t, unsigned int, -1, -1, -1, true);
+      }
+    }
+  } else {
+    cuda::detail::TensorInfo<scalar_t, uint64_t> outInfo =
+      cuda::detail::getTensorInfo<scalar_t, uint64_t>(out, /*positiveDim = */ true);
+    int outSelectDim = outInfo.collapseDims(dim);
+    outInfo.reduceDim(outSelectDim);
+
+    cuda::detail::TensorInfo<scalar_t, uint64_t> selfInfo = cuda::detail::getTensorInfo<scalar_t, uint64_t>(self, /*positiveDim = */ true);
+    int selfSelectDim = selfInfo.collapseDims(dim);
+    selfInfo.reduceDim(selfSelectDim);
+
+    cuda::detail::TensorInfo<int64_t, uint64_t> indicesInfo = cuda::detail::getTensorInfo<int64_t, uint64_t>(index, /*positiveDim = */ true);
+    indicesInfo.collapseDims();
+
+    LARGE_INDEX(scalar_t, uint64_t, -1, -1, -1, true);
+  }
+#undef SMALL_INDEX
+#undef LARGE_INDEX
+}
+} // anonymous namespace
+
+Tensor& index_select_out_cuda(Tensor& out, const Tensor& self, long dim,
+                              const Tensor& index) {
+  static constexpr int MAX_INDEX_SELECT_DIMS = 25;
+  static constexpr string_view DIM_WARNING =
+    "Tensor too large or too many (> 25) dimensions";
+
+  TORCH_CHECK(out.device() == self.device(), "Input and output must be on the "
+              "same device");
+
+  dim = at::maybe_wrap_dim(dim, self);
+  TORCH_CHECK(out.dim() <= MAX_INDEX_SELECT_DIMS, DIM_WARNING);
+  TORCH_CHECK(self.dim() <= MAX_INDEX_SELECT_DIMS, DIM_WARNING);
+  TORCH_CHECK(index.dim() <= MAX_INDEX_SELECT_DIMS, DIM_WARNING);
```

Comment:
This comes from `MAX_TENSORINFO_DIMS`, right? Maybe use that macro constant directly instead of the magic number 25.